pallets/gear/src/lib.rs

```diff
@@ -1242,9 +1243,27 @@ pub mod pallet {
 
             let (builtins, _) = T::BuiltinDispatcherFactory::create();
             let ext_manager = ExtManager::<T>::new(builtins);
-            ext_manager.set_program(packet.destination(), &code_info, message_id, block_number);
 
             let program_id = packet.destination();
+
+            // Before storing the program to `ProgramStorage` we need to make sure that an account
+            // can be created for the program.
+            // Note: making a transfer outside of the `Ext::set_program()` because here a transfer
+            // is allowed to fail (as opposed to creating a program by a program).
+            CurrencyOf::<T>::transfer(
+                &who,
+                &program_id.cast(),
+                CurrencyOf::<T>::minimum_balance(),
+                ExistenceRequirement::KeepAlive,
+            )?;
+
+            // A consumer is added at program's account creation and removed it on termination
+            // to make sure the account is not removed by the runtime while the program exists.
+            // With a consumer, a correct runtime should not remove the account.
+            System::<T>::inc_consumers(&program_id.cast())?;
```

Comment:
you need to lock this funds, that automatically will inc consumers/providers

Reply:
> you need to lock this funds, that automatically will inc consumers/providers
Can't see anything like that in Substrate code.

Anyway, having funds locked doesn't necessarily exclude them from reducible amount whereas having non-zero consumers does.
That's how the `Contracts` or `Assets` pallets go about it - by explicitly incrementing consumers for a contract's account to enforce the same invariant of the account to always exist.
Locks serve a different purpose.

Reply:
Again, the latter argument is no longer valid and can be ignored. Locks are used now.